docs/decisions/0002-feature-based-application-organization.rst

```diff
@@ -0,0 +1,213 @@
+      /service.js // Note: singular - there's one 'service' here that provides many methods.
+
+If you find yourself desiring to have multiple files of a particular type in the data directory, this is a strong sign that you actually need a sub-feature instead.
+
+**C. Sub-features follow the same naming scheme.**
+
+Sub-features should follow the same rules as any other module.
+
+A module with a sub-module:
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx
+    /Avatar.jsx // additional components for a feature reside here at the top level, not in a "components" subdirectory.
+    /data
+      /actions.js
+      /reducers.js
+      /sagas.js
+      /service.js
+    /profile-photo
+      /index.js // public interface
+      /ProfilePhoto.jsx
+      /data
+        /actions.js
+        /reducers.js
+        /selectors.js
+    /education // Sparse sub-module
+      /index.js // public interface
+      /Education.jsx
+    /site-language // No view layer sub-module
+      /index.js // public interface
+      /data
+        /actions.js
+        /reducers.js
+
+Note that a given feature need not contain files of all types, nor is having files of all types a prerequisite for having a feature. A feature may not contain a view (Component) layer, or in contrast to that, may not need a data directory at all!
+
+Importing rules of thumb
+========================
+
+It can be difficult to figure out where it's okay to import from. Following these rules of thumb will help maintain a healthy code organization and should prevent the possibility of circular dependencies.
+
+**I. A feature may not import from its parentage.**
+
+As described above in "Avoid circular dependencies", features should not import from their parent, grandparent, etc. A feature should be agnostic to the context in which it is used. If a module is importing from its parent or grandparent, that implies something is ill-factored.
+
+**II. A feature may import from its children, but not its grandchildren.**
+
+The feature may only import from the exports of its child, which may include exports of the grandchildren. Importing directly from grandchildren (or great grandchildren, etc.) would violate the strict module boundary of the child.
+
+**II. Features may import from their siblings.**
+
+It's acceptable to import from a module's siblings, or the siblings of their parents, grandparents, etc. This is necessary to support code re-use. As an example, assume we have a sub-module with common code to support our web forms.
+
+::
+
+  /feature1
+    /sub-form-1
+    /sub-form-2
+    /forms-common-code
+
+The sub-form modules can import from forms-common-code. The latter has its own strict module boundary and could conceptually be extracted into its own repository/completely independent module as far as they're concerned. They're unaware, conceptually, that it's a child of feature1, and they don't care.
+
+**III. Features may import from the siblings of their parentage.**
```

Comment:
I understand the necessity of this, but it does break the "A feature should be agnostic to the context in which it is used" principle mentioned earlier; e.g., if `forms-common-code` is moved another directory level higher, it will break all imports of it.

Python recommends handling this by doing absolute project-level imports for common code (`from project.common import blah`). I'm not as experienced with front-end development, though, so I don't know if this translates well.

Reply:
Truth.  I think this is where the 'spirit' vs. the exact letter of the rule comes into play, as we're somewhat limited by the way the import statements are written.  Webpack does give us a way of resolving imports from an absolute directory, but it feels a bit worse to me to rely on that tool's feature rather than letting the imports be short and simple for the most part.  

If a module gets extracted into an actual module in node_modules, exactly what you describe will happen, which is great.  But in the meantime, it feels acceptable for modules that are only in a single app to use relative paths to help route themselves.  That's not terribly defensible, but it feels okay from having seen it in various apps. :)